File: apps/ihg/src/server/entity/Schedule.ts
import { InferSchemaType, model, models, Schema } from "mongoose";

const ScheduleSchema = new Schema({
  sport: {
    type: String,
    required: true,
  },
  halls: [
    {
      name: {
        type: String,
        required: true,
      },
      imgUrl: {
        type: String,
        required: true,
      },
      colourCode: {
        type: String,
        required: true,
      },
      score: {
        type: Number,
      },
      abbreviation: {
        type: String,
        required: true,
      },
    },
  ],
  // get date from here
  startTime: {
    type: Date,
    required: true,
  },
  endTime: {
    type: Date,
    required: true,
  },
  venue: {
    type: String,
    required: true,
  },
  gender: {
    type: String,
    enum: ["Male", "Female", "Mixed"],
    required: true,
  },
  stage: {
    type: String,
    enum: [
      "Group A",
      "Group B",
      "Prelims",
      "Semi 1",
      "Semi 2",
      "Finals",
      "Carnival",
      "Playoffs",
    ],
    required: true,
  },
});

export type ScheduleDocument = InferSchemaType<typeof ScheduleSchema> & {
  _id: string;
};

const Schedule = models.Schedule || model("Schedule", ScheduleSchema);

export default Schedule;

File: pkg/models/schedule.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

// ScheduleHall mirrors the halls sub-document embedded in a Schedule.
type ScheduleHall struct {
	Name         string   `bson:"name" json:"name"`
	ImgUrl       string   `bson:"imgUrl" json:"imgUrl"`
	ColourCode   string   `bson:"colourCode" json:"colourCode"`
	Score        *float64 `bson:"score,omitempty" json:"score,omitempty"`
	Abbreviation string   `bson:"abbreviation" json:"abbreviation"`
}

// Gender enum values allowed on a Schedule, mirroring the Mongoose schema enum.
var ValidGenders = map[string]bool{"Male": true, "Female": true, "Mixed": true}

// Stage enum values allowed on a Schedule, mirroring the Mongoose schema enum.
var ValidStages = map[string]bool{
	"Group A": true, "Group B": true, "Prelims": true,
	"Semi 1": true, "Semi 2": true, "Finals": true,
	"Carnival": true, "Playoffs": true,
}

type Schedule struct {
	ID        bson.ObjectID  `bson:"_id,omitempty" json:"_id,omitempty"`
	Sport     string         `bson:"sport" json:"sport"`
	Halls     []ScheduleHall `bson:"halls" json:"halls"`
	StartTime time.Time      `bson:"startTime" json:"startTime"`
	EndTime   time.Time      `bson:"endTime" json:"endTime"`
	Venue     string         `bson:"venue" json:"venue"`
	Gender    string         `bson:"gender" json:"gender"`
	Stage     string         `bson:"stage" json:"stage"`
}
